test/fnn/test_fnn_cpu.cpp

```diff
@@ -0,0 +1,193 @@
+
+bool test_fnn_cpu(bool recompute_outputs, std::string date, std::string arch,
+                  std::string data) {
+    // Create TAGI network
+    Network net;
+
+    net.layers = LAYERS;
+    if (data == "Boston_housing") {
+        net.nodes = NODES_BH;
+    } else {
+        net.nodes = NODES_1D;
+        net.activations = ACTIVATIONS;
+    }
+    net.activations = ACTIVATIONS;
+    net.batch_size = BATCH_SIZE;
+    net.sigma_v = 0.06;
+    net.sigma_v_min = 0.06;
+
+    TagiNetworkCPU tagi_net(net);
+
+    // Put it in the main test file
+    SavePath path;
+    path.curr_path = get_current_dir();
+    std::string data_path = path.curr_path + "/test/data/" + data;
+    std::string init_param_path_w = path.curr_path + "/test/" + arch +
+                                    "/data/" + date + "_init_param_weights_" +
+                                    arch + "_" + data + ".csv";
+    std::string init_param_path_b = path.curr_path + "/test/" + arch +
+                                    "/data/" + date + "_init_param_bias_" +
+                                    arch + "_" + data + ".csv";
+    std::string opt_param_path_w = path.curr_path + "/test/" + arch + "/data/" +
+                                   date + "_opt_param_weights_" + arch + "_" +
+                                   data + ".csv";
+    std::string opt_param_path_b = path.curr_path + "/test/" + arch + "/data/" +
+                                   date + "_opt_param_bias_" + arch + "_" +
+                                   data + ".csv";
+    std::string forward_states_path =
+        path.curr_path + "/test/" + arch + "/data/" + date +
+        "_forward_hidden_states_" + arch + "_" + data + ".csv";
+    std::string forward_states_path_2 =
+        path.curr_path + "/test/" + arch + "/data/" + date +
+        "_forward_hidden_states_2_" + arch + "_" + data + ".csv";
+    std::string backward_states_path =
+        path.curr_path + "/test/" + arch + "/data/" + date +
+        "_backward_hidden_states_" + arch + "_" + data + ".csv";
+    std::string backward_states_path_2 =
+        path.curr_path + "/test/" + arch + "/data/" + date +
+        "_backward_hidden_states_2_" + arch + "_" + data + ".csv";
+
+    // Train data
+    Dataloader train_db = train_data(data, tagi_net, data_path, NORMALIZE);
+    // Test data
+    Dataloader test_db =
+        test_data(data, tagi_net, data_path, train_db, NORMALIZE);
+
+    std::vector<std::vector<float> *> weights;
+    weights.push_back(&tagi_net.theta.mw);
+    weights.push_back(&tagi_net.theta.Sw);
+    weights.push_back(&tagi_net.theta.mb);
+    weights.push_back(&tagi_net.theta.Sb);
+
+    std::vector<std::vector<float> *> bias;
+    bias.push_back(&tagi_net.theta.mw_sc);
+    bias.push_back(&tagi_net.theta.Sw_sc);
+    bias.push_back(&tagi_net.theta.mb_sc);
+    bias.push_back(&tagi_net.theta.Sb_sc);
+
+    if (recompute_outputs) {
+        write_vector_to_csv(init_param_path_w, "mw,Sw,mb,Sb", weights);
```

Comment:
@miquelflorensa I have an error message when running `build/main test -reset all`. This line gave me that error. It won't be able to write a file if the file does not exist yer. You can reproduce it by clearing out all data regarding the csv file relating to the parameters and run the command

Reply:
I woud write `mw & Sw`,  `mb & Sb`, `mw_sc & Sw_sc` and   `mb_sc & Sb_sc` in 4 different files because their size are different from one to another. Note that w: weights, b: bias, w_sc: short-cut weight, b_sc: short-cut bias